src/deepsparse/transformers/pipelines/text_generation.py

```diff
@@ -443,28 +443,64 @@ def engine_inputs_for_prefill(
         self, tokens: List[int]
     ) -> Generator[List[numpy.ndarray], None, None]:
         """
-        Takes a list of tokens and turns the first
-        `self.prompt_processing_sequence_length` tokens into
-        appropriate engine inputs for the multitoken engine.
+        Takes a list of tokens and creates a generator
+        of engine_inputs for the multitoken engine.
+
+        The tokens first get batched into chunks that
+        fit the expected input_ids length of the multitoken
+        engine.
+
+        Then, from every batch of tokens an appriopriate
+        engine_inputs data structure is created.
 
         :param tokens: the list of tokens to process
         :return: a generator of engine inputs
         """
-        # TODO: Make it yield multiple engine_inputs
-        # Once this is done, we can update the docstrings and be
-        # more verbose about what this function does
-        engine_inputs = []
-        token_batches = [tokens[: self.prompt_processing_sequence_length]]
-        for token_batch in token_batches:
+
+        num_batches = len(tokens) // self.prompt_processing_sequence_length
+
+        token_batches = [
+            tokens[i : i + self.prompt_processing_sequence_length]
+            for i in range(num_batches)
+        ]
+
+        for idx, token_batch in enumerate(token_batches):
+
+            engine_inputs = []
+
+            # fetch the number of non-blank cache entries
+            num_non_blank_cache_entries = (
+                self.multitoken_engine.num_non_blank_cache_entries
+            )
+            # if there are more non-blank entries than the
+            # full size of the attention mask, trim to attention
+            # mask length (self.sequence_length, but we will also
+            # be adding contribution from
+            # self.prompt_processing_sequence_length)
+            num_non_blank_cache_entries = min(
+                num_non_blank_cache_entries,
+                self.sequence_length - self.prompt_processing_sequence_length,
```

Comment:
shouldn't this be the total remaining tokens ie something like ` self.sequence_length - idx * self.prompt_processing_sequence_length` or am I missing something?

Reply:
We are essentially talking about the same thing, but my logic was way too overcomplicated. Refactored the function, so now hopefully anyone reading should grasp what's going on.